dkeyserver/Enclave/enclave.cpp

```diff
@@ -494,3 +494,278 @@ int sgx_set_up_tls_server(char *server_port)
+{
+    log_d(TLS_SERVER "Connecting target server...\n");
+
+    int ret = 0;
+
+    SSL_CTX *ssl_client_ctx = nullptr;
+    SSL *ssl_session = nullptr;
+
+    X509 *cert = nullptr;
+    EVP_PKEY *pkey = nullptr;
+    SSL_CONF_CTX *ssl_confctx = SSL_CONF_CTX_new();
+
+    int client_socket = -1;
+    int error = 0;
+    if (target_server_name == NULL)
+    {
+        log_d(TLS_SERVER "Connecting target failed: target server name unavailable.\n");
+        goto done;
+    }
+
+    if ((ssl_client_ctx = SSL_CTX_new(TLS_client_method())) == nullptr)
+    {
+        log_d(TLS_SERVER "Unable to create a new SSL context\n");
+        goto done;
+    }
+
+    if (initalize_ssl_context(ssl_confctx, ssl_client_ctx) != SGX_SUCCESS)
+    {
+        log_d(TLS_SERVER "Unable to create a initialize SSL context\n ");
+        goto done;
+    }
+
+    // specify the verify_callback for verification
+    SSL_CTX_set_verify(ssl_client_ctx, SSL_VERIFY_PEER, &verify_callback);
+    log_d(TLS_SERVER "Load cert and key\n");
+    if (load_tls_certificates_and_keys(ssl_client_ctx, cert, pkey) != 0)
+    {
+        log_d(TLS_SERVER "Unable to load certificate and private key on the client\n");
+        goto done;
+    }
+
+    if ((ssl_session = SSL_new(ssl_client_ctx)) == nullptr)
+    {
+        log_d(TLS_SERVER "Unable to create a new SSL connection state object\n");
+        goto done;
+    }
+
+    log_d(TLS_SERVER "New ssl connection getting created\n");
+    client_socket = create_socket(target_server_name, target_server_port);
+    if (client_socket == -1)
+    {
+        log_d(
+            TLS_SERVER "Create a socket and initiate a TCP connect to target server: %s:%d "
+            "(errno=%d)\n",
+            target_server_name,
+            target_server_port,
+            errno);
+        goto done;
+    }
+
+    // set up ssl socket and initiate TLS connection with TLS target server
+    if (SSL_set_fd(ssl_session, client_socket) != 1)
+    {
+        log_d(TLS_SERVER "Ssl set fd error.\n");
+    }
+    else
+    {
+        log_d(TLS_SERVER "Ssl set fd succeed.\n");
+    }
+
+    if ((error = SSL_connect(ssl_session)) != 1)
+    {
+        log_d(TLS_SERVER "Error: Could not establish a TLS session ret2=%d "
+                       "SSL_get_error()=%d\n",
+            error,
+            SSL_get_error(ssl_session, error));
+        goto done;
+    }
+    log_d(TLS_SERVER "successfully established TLS channel:%s\n",
+        SSL_get_version(ssl_session));
+
+    // start the communication
+    if ((error = get_dkey_from_target(ssl_session)) != 0)
+    {
+        log_d(TLS_SERVER "Failed: get_dkey_from_target (ret=%d)\n", error);
+        goto done;
+    }
+
+    // Free the structures we don't need anymore
+    ret = 0;
+done:
+
+    if (client_socket != -1)
+    {
+        ocall_close(&ret, client_socket);
+        if (ret != 0)
+            log_d(TLS_SERVER "OCALL: error close socket\n");
```

Comment:
if fail at line 727, ret is -1, here ocall_close ret is 0 and return as success.

Reply:
done.